realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1157,6 +1160,153 @@ public void run() {
+                if (Thread.currentThread().isInterrupted()) {
+                    return;
+                }
+
+                boolean transactionCommitted = false;
+                final Throwable[] exception = new Throwable[1];
+                final Realm bgRealm = Realm.getInstance(realmConfiguration);
+                bgRealm.beginTransaction();
+                try {
+                    transaction.execute(bgRealm);
+
+                    if (!Thread.currentThread().isInterrupted()) {
+                        bgRealm.commitTransaction(new Runnable() {
+                            @Override
+                            public void run() {
+                                // The bgRealm needs to be closed before post event to caller's handler to avoid
+                                // concurrency problem. eg.: User wants to delete Realm in the callbacks.
+                                // This will close Realm before sending REALM_CHANGED.
+                                bgRealm.close();
+                            }
+                        });
+                        transactionCommitted = true;
+                    }
+                } catch (final Throwable e) {
+                    exception[0] = e;
+                } finally {
+                    if (!bgRealm.isClosed()) {
+                        if (bgRealm.isInTransaction()) {
+                            bgRealm.cancelTransaction();
+                        } else if (exception[0] != null) {
+                            RealmLog.w("Could not cancel transaction, not currently in a transaction.");
+                        }
+                        bgRealm.close();
+                    }
+
+                    // Send response as the final step to ensure the bg thread quit before others get the response!
+                    if (handler != null
+                            && !Thread.currentThread().isInterrupted()
+                            && handler.getLooper().getThread().isAlive()) {
+                        if (onSuccess != null && transactionCommitted) {
+                            handler.post(new Runnable() {
+                                @Override
+                                public void run() {
+                                    onSuccess.onSuccess();
+                                }
+                            });
+                        }
+                        final Throwable backgroundException = exception[0];
+                        if (backgroundException != null) {
+                            if (onError != null) {
+                                handler.post(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        onError.onError(backgroundException);
+                                    }
+                                });
+                            } else {
+                                handler.post(new Runnable() {
```

Comment:
There isn't any difference between this and triggering the error callback above right? I guess that technically you can quit a looper thread before it finished receiving all events, but by that assumption we couldn't propagate any errors. I would probably say that chance of this happening is so little, and it would be against the contract you agreed to when doing it async, so basically "it is your own fault"?